FIAT/reference_element.py

```diff
@@ -507,11 +507,157 @@ def construct_subelement(self, dimension):
+        reference triangle,
+            `P0 = (0, 0)`,
+            `P1 = (1, 0)` and
+            `P2 = (0, 1)`.
+        Our point is
+            `X = [x, y]`.
+        Below is a diagram of the cell (which may not render correctly in
+        sphinx):
+
+        .. code-block:: text
+        ```
+                y-axis
+                |
+                |
+          (0,1) P2
+                | \\
+                |  \\
+                |   \\
+                |    \\
+                |  T  \\
+                |      \\
+                |       \\
+                |        \\
+            ---P0--------P1--- x-axis
+          (0,0) |         (1,0)
+        ```
+
+        Barycentric coordinates are defined as
+            `X = alpha * P0 + beta * P1 + gamma * P2` where
+            `alpha + beta + gamma = 1.0`.
+        The solution is
+            `alpha = 1 - X[0] - X[1] = 1 - x - y`,
+            `beta = X[0] = x` and
+            `gamma = X[1] = y`.
+        If all three are positive, the point is inside the reference cell.
+        If any are negative, we are outside it. The absolute sum of any
+        negative barycentric coordinates usefully gives the L1 distance from
+        the cell to the point. For example the point (1,1) has L1 distance
+        1 from the cell: on this case alpha = -1, beta = 1 and gamma = 1.
+        -alpha = 1 is the L1 distance. For comparison the L2 distance (the
+        length of the vector from the nearest point on the cell to the point)
+        is sqrt(0.5^2 + 0.5^2) = 0.707. Similarly the point (-1.0, -1.0) has
+        alpha = 3, beta = -1 and gamma = -1. The absolute sum of beta and gamma
+        2 which is again the L1 distance. The L2 distance in this case is
+        sqrt(1^2 + 1^2) = 1.414.
+
+        For a UFCTetrahedron we have four vertices
+            `P0 = (0,0,0)`,
+            `P1 = (1,0,0)`,
+            `P2 = (0,1,0)` and
+            `P3 = (0,0,1)`.
+        Our point is
+            `X = [x, y, z]`.
+        The barycentric coordinates are defined as
+            `X = alpha * P0 + beta * P1 + gamma * P2 + delta * P3`
+            where
+            `alpha + beta + gamma + delta = 1.0`.
+        The solution is
+            `alpha = 1 - X[0] - X[1] - X[2] = 1 - x - y - z`,
+            `beta = X[0] = x`,
+            `gamma = X[1] = y` and
+            `delta = X[2] = z`.
+        The rules are the same as for the triangle but with one extra
+        barycentric coordinate. Our approximate distance, the absolute sum of
+        the negative barycentric coordinates, is at worse around 4 times the
+        actual distance to the tetrahedron.
+
+        """
+        # bary = [alpha, beta, gamma, delta, ...] - see docstring
+        bary = [1.0 - sum(point)] + list(point)
+        # We output the most negative of these in a way the sympy can cope
+        # with. bary-abs(bary) gets rid of the positive barycentric coordinates
```

Comment:
Say "we avoid branching so that code can be generated from this (e.g. with sympy)"